File: flutter/packages/workout_log_core/lib/src/models/session.dart
import '../coding.dart';
import 'block.dart';

/// Files on disk are the source of truth (ADR-001); this is a lossless projection.
class Session with JsonEquality {
  Session({
    required this.date,
    required this.cycleDay,
    this.startTime,
    this.kind = Kind.training,
    this.bodyweightKg,
    this.notes,
    List<Block>? blocks,
  }) : blocks = blocks ?? [];

  factory Session.fromJson(Map<String, dynamic> json) => Session(
        date: json['date'] as String,
        cycleDay: json['cycle_day'] as String,
        startTime: json['start_time'] as String?,
        kind: Kind.fromJson(json['kind'] as String?) ?? Kind.training,
        bodyweightKg: asDouble(json['bodyweight_kg']),
        notes: json['notes'] as String?,
        blocks: [
          for (final b in json['blocks'] as List)
            Block.fromJson(b as Map<String, dynamic>),
        ],
      );

  String date;
  String cycleDay;
  String? startTime;
  Kind kind;
  double? bodyweightKg;
  String? notes;
  List<Block> blocks;

  @override
  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{'date': date, 'cycle_day': cycleDay};
    put(json, 'start_time', startTime);
    json['kind'] = kind.name;
    put(json, 'bodyweight_kg', bodyweightKg);
    put(json, 'notes', notes);
    json['blocks'] = [for (final b in blocks) b.toJson()];
    return json;
  }

  Session copy() => Session.fromJson(toJson());
}

enum Kind {
  training,
  deload,
  retest;

  static Kind? fromJson(String? value) =>
      value == null ? null : Kind.values.byName(value);
}

File: crates/workout-log-core/src/models/session.rs
use serde::{Deserialize, Serialize};

use super::block::Block;

/// Files on disk are the source of truth (ADR-001); this is a lossless projection.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Session {
    pub date: String,
    pub cycle_day: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub start_time: Option<String>,
    #[serde(default)]
    pub kind: Kind,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bodyweight_kg: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub notes: Option<String>,
    #[serde(default)]
    pub blocks: Vec<Block>,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Kind {
    #[default]
    Training,
    Deload,
    Retest,
}

impl Kind {
    pub const ALL: [Self; 3] = [Self::Training, Self::Deload, Self::Retest];
}
